### PR TITLE
Check CSV Upload permissions on the frontend

### PR DESCRIPTION
**Pull request by iethree**
Closes https://github.com/metabase/metabase/issues/30111

### Description

This should properly show CSV upload UI based on a user's permissions and instance settings.

Namely: a user can upload a CSV if:
1. Uploads are enabled and configured on the instance, and
2. The user has "curate" permissions on the collection they're uploading to, AND
3. The user has data access to the database they're uploading to

User Cannot Upload | User Can Upload
---|---
Screen Shot 2023-04-13 at 2 34 55 PM | Screen Shot 2023-04-13 at 2 35 02 PM

### How to verify

- In metabase enterprise edition, set up a user who has view-only access to a collection
- see that the limited user cannot upload to that collection
- Turn on/off uploads in admin settings and see that the upload button and dropzone get enabled/disabled appropriately for admin users

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30094)

### CODE REVIEW

File: src/metabase/public_settings.clj [hunk 1]
```diff
@@ -602,21 +602,21 @@
...
   (deferred-tru "Whether or not uploads are enabled")
-  :visibility :admin
+  :visibility :authenticated
   :type       :boolean
   :default    false)
 
 (defsetting uploads-database-id
   (deferred-tru "Database ID for uploads")
-  :visibility :admin
+  :visibility :authenticated
```
[the remark this answers is not in the document]
Comment (iethree) [reply src/metabase/public_settings.clj:611]: even if uploads are enabled via toggle, if other options are not set properly, a user won't be able to actually upload a document. The FE needs to check if these options are non-empty to enable/disable the upload UI.

**Review by alxnddr — COMMENTED** [review]
Should we also check if a user has the access to the DB?

**Comment by iethree** [comment]
> [](https://metaboat.slack.com/archives/C04S696LRUM/p1681418036100589)

On third thought, I asked some more questions and this probably needs to be refined to make sure that anything you can upload you can also view. 🤔

**Review by alxnddr — COMMENTED** [review]
Looks great! Here are a couple of things I found while looking at the branch:

1) If a user has "no self-service" permission to the uploads DB the collection view looks like this:
Screen Shot 2023-04-18 at 6 33 00 PM

2) Seems like there is no check on the BE since it lets me upload CSVs even if I have no self-service access level.

3) The uploading state behavior seems not fully working

<video src="https://user-images.githubusercontent.com/14301985/232911828-4384c8a6-c613-44e6-9e65-a55c58f8148f.mov" />

File: frontend/src/metabase/status/components/StatusLarge/StatusLarge.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  StatusToggle,
+  StatusBody,
+} from "./StatusLarge.styled";
+
+type status = {
+  title: string;
+  items: statusItem[];
+};
+
+type statusItem = {
```
Comment (alxnddr) [inline frontend/src/metabase/status/components/StatusLarge/StatusLarge.tsx:25]: Pascal case for types is used in other places, wdyt about sticking to it? Also, these could be interfaces

File: frontend/src/metabase/admin/settings/components/widgets/UploadSettingsWidget/UploadSettingsWidget.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import type { Database as DatabaseType } from "metabase-types/types/Database";
+
+import SettingSelect from "../SettingSelect";
+
+const getDatabaseOptions = (databases: DatabaseType[]) =>
+  databases
+    .filter(db => db?.settings?.["database-enable-actions"])
+    .map(db => ({ name: db.name, value: db.id }));
+
+// const enginesWithSchema = ["postgres"];
```
Comment (alxnddr) [inline frontend/src/metabase/admin/settings/components/widgets/UploadSettingsWidget/UploadSettingsWidget.tsx:13]: Some commented code slipped here

File: frontend/src/metabase/collections/components/CollectionHeader/CollectionUpload.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+const CLEAR_ERROR_TIMEOUT = 3000;
+
+export default function ColllectionUpload({
+  collection,
+  onUpload,
+}: {
+  collection: Collection;
+  onUpload: (file: File, collectionId: CollectionId) => void;
+}) {
+  const [error, setError] = useState("");
```
Comment (alxnddr) [inline frontend/src/metabase/collections/components/CollectionHeader/CollectionUpload.tsx:18]: Where `error` is being set?

File: frontend/src/metabase/lib/uploads.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { FileUpload } from "metabase-types/store/upload";
+
+export const uploadInProgress = (upload: FileUpload) =>
```
Comment (alxnddr) [inline frontend/src/metabase/lib/uploads.ts:3]: nit: in most places, we prefer using `isUploadInProgress` over `uploadInProgress` for predicates

File: frontend/src/metabase/collections/components/CollectionHeader/CollectionUpload.tsx (hunk 4 above)
Comment (iethree) [reply frontend/src/metabase/collections/components/CollectionHeader/CollectionUpload.tsx:18]: ah good catch. there's a bunch of code here that became defunct when the notification got put in.

**Comment by iethree** [comment]
@alxnddr thanks again for the comments and video. I've addressed all of them except the second-upload-doesnt-trigger-status-popup one. I couldn't reproduce that. Maybe I accidentally fixed that with some other stuff?
multiple-uploads

File: frontend/src/metabase/status/components/StatusListing/StatusListing.tsx [hunk 6]
```diff
@@ -1,22 +1,17 @@
...
 
-const StatusListing = ({
-  isAdmin,
-  database: uploadDatabase,
-}: StatusListingProps): JSX.Element => {
+const StatusListing = ({ isAdmin }: StatusListingProps): JSX.Element => {
   return (
     <StatusListingRoot>
       {isAdmin && <DatabaseStatus />}
-      {uploadDatabase?.canWrite() && <FileUploadStatus />}
```
Comment (iethree) [inline frontend/src/metabase/status/components/StatusListing/StatusListing.tsx:19]: @npfitz I don't see any reason to check authentication here at all. I think the right place to check permissions is whether the user can upload at all. if that is properly handled, they'll never see this status indicator at all, right?

In any event `.canWrite()` I think means "has sql query permissions", which doesn't quite align with "can upload CSVs" right now.

**Review by alxnddr — APPROVED** [review]

**Pull request merged**